editor/src/messages/layout/layout_message_handler.rs

```diff
@@ -208,55 +239,57 @@ impl<F: Fn(&MessageDiscriminant) -> Vec<KeysGroup>> MessageHandler<LayoutMessage
 }
 
 impl LayoutMessageHandler {
+	/// Diff the update and send to the frontend where necessary
+	fn send_layout(&mut self, layout_target: LayoutTarget, new_layout: Layout, responses: &mut VecDeque<Message>, action_input_mapping: &impl Fn(&MessageDiscriminant) -> Vec<KeysGroup>) {
+		// We don't diff the menu bar layout yet.
+		if layout_target == LayoutTarget::MenuBar {
+			// Skip update if the same
+			if self.layouts[layout_target as usize] == new_layout {
+				return;
+			}
+			// Update the backend storage
+			self.layouts[layout_target as usize] = new_layout;
+			// Update the UI
+			responses.push_back(
+				FrontendMessage::UpdateMenuBarLayout {
+					layout_target,
+					layout: self.layouts[layout_target as usize].clone().unwrap_menu_layout(action_input_mapping).layout,
+				}
+				.into(),
+			);
+			return;
+		}
+
+		let mut widget_diffs = Vec::new();
+		self.layouts[layout_target as usize].diff(new_layout, &mut Vec::new(), &mut widget_diffs);
+		// Skip sending if no diff.
+		if widget_diffs.is_empty() {
+			return;
+		}
+
+		self.send_diff(widget_diffs, layout_target, responses, action_input_mapping);
+	}
+
+	/// Send a diff to the frontend based on the layout target.
 	#[remain::check]
-	fn send_layout(&self, layout_target: LayoutTarget, responses: &mut VecDeque<Message>, action_input_mapping: &impl Fn(&MessageDiscriminant) -> Vec<KeysGroup>) {
-		let layout = &self.layouts[layout_target as usize];
+	fn send_diff(&self, mut diff: Vec<WidgetDiff>, layout_target: LayoutTarget, responses: &mut VecDeque<Message>, action_input_mapping: &impl Fn(&MessageDiscriminant) -> Vec<KeysGroup>) {
+		diff.iter_mut().for_each(|diff| diff.new_val.apply_shortcut(action_input_mapping));
+
+		info!("{layout_target:?} diff {diff:#?}");
```

Comment:
Please remove this log, or switch it to `trace!()`. In the future I also recommend debug logging with `debug!()` which, just like `info!()`, is always enabled regardless of debug verbosity level, and it's easier to remember that we are supposed to remove `debug!()` statements before committing. We use `info!()` for things like printing the message tree or serialized document format, when the user specifically requests something be shown and the log verbosity level is irrelevant in that. Thanks :)

Reply:
`debug!` is too many keystrokes :/.

Reply:
Haha, fair enough :)

Reply:
And at least now I'll probably remember exactly what `info!()` is so I can keep an eye out during code review for that.